prow/sidecar/censor.go

```diff
@@ -112,6 +116,218 @@ func (o Options) censor() error {
+	}
+	zipReader, err := gzip.NewReader(input)
+	if err != nil {
+		return fmt.Errorf("could not read archive: %w", err)
+	}
+	tarReader := tar.NewReader(zipReader)
+	defer func() {
+		if err := zipReader.Close(); err != nil {
+			logrus.WithError(err).Warn("Could not close zip reader after unarchiving.")
+		}
+		if err := input.Close(); err != nil {
+			logrus.WithError(err).Warn("Could not close input file after unarchiving.")
+		}
+	}()
+
+	for {
+		entry, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("could not read archive: %w", err)
+		}
+		if !validRelPath(entry.Name) {
+			return fmt.Errorf("tar contained invalid name error %q", entry.Name)
+		}
+		rel := filepath.FromSlash(entry.Name)
+		abs := filepath.Join(destPath, rel)
+		mode := entry.FileInfo().Mode()
+		switch {
+		case mode.IsDir():
+			if err := os.MkdirAll(abs, 0755); err != nil {
+				return fmt.Errorf("could not create directory while unpacking archive: %w", err)
+			}
+		case mode.IsRegular():
+			file, err := os.OpenFile(abs, os.O_RDWR|os.O_CREATE|os.O_TRUNC, mode.Perm())
+			if err != nil {
+				return err
+			}
+			n, err := io.Copy(file, tarReader)
+			if closeErr := file.Close(); closeErr != nil && err == nil {
+				return fmt.Errorf("error closing %s: %v", abs, closeErr)
+			}
+			if err != nil {
+				return fmt.Errorf("error writing to %s: %v", abs, err)
+			}
+			if n != entry.Size {
+				return fmt.Errorf("only wrote %d bytes to %s; expected %d", n, abs, entry.Size)
+			}
+		}
+	}
+	return nil
+}
+
+func validRelPath(p string) bool {
+	if p == "" || strings.Contains(p, `\`) || strings.HasPrefix(p, "/") || strings.Contains(p, "../") {
+		return false
+	}
+	return true
+}
+
+// archive re-packs the dir into the destination
+func archive(srcDir, destArchive string) error {
+	output, err := ioutil.TempFile("", "tmp-archive")
+	if err != nil {
+		return fmt.Errorf("failed to create temporary file for archive: %w", err)
+	}
+
+	zipWriter := gzip.NewWriter(output)
+	tarWriter := tar.NewWriter(zipWriter)
+	defer func() {
+		if err := tarWriter.Close(); err != nil {
+			logrus.WithError(err).Warn("Could not close tar writer after archiving.")
+		}
+		if err := zipWriter.Close(); err != nil {
+			logrus.WithError(err).Warn("Could not close zip writer after archiving.")
+		}
+		if err := output.Close(); err != nil {
+			logrus.WithError(err).Warn("Could not close output file after archiving.")
+		}
+	}()
+
+	if err := filepath.Walk(srcDir, func(absPath string, info os.FileInfo, err error) error {
+		header, err := tar.FileInfoHeader(info, info.Name())
+		if err != nil {
+			return fmt.Errorf("could not create tar header: %w", err)
+		}
+		// the header won't get nested paths right
+		relpath, _ := filepath.Rel(srcDir, absPath) // err happens when there's no rel path, but we know there must be
+		header.Name = relpath
+		// in order to get reproducible output archives, we need to remove variance
```

Comment:
IMHO this section is not contentious, but it *does* change the archive as given by the user. I claim it's not contentious as the modification time, user bits etc for an artifact are not really that interesting. Who cares if `g+w` on the file in GCS?

Reply:
yeah doesn't really matter but why does this need to be reproducible? I think for tests it would be easiest to just let `tar` create an archive from some data we provide, then use that. That most closely mirrors reality IMHO

Reply:
So only test that `archive()` creates an archive that `tar` can extract into the original input, but test nothing about the actual shape of the archive?

Reply:
What is important here is IMHO a) We can cope with archives created by tar b) tar can cope with archives created by us. I don't think it is super important to check that the archive we create looks the same as one tar creates plus such a test might be pretty brittle and e.G. not work on different tar versions because some default changed or whatnot

Reply:
Sounds good to me